### ISSUE TITLE
Implicit provider inheritance not working as expected. 

### ISSUE THREAD

**Issue by dsfrederic** [hashicorp/terraform#28157]
### Terraform Version

```
0.14.8
```

### Terraform Configuration Files

```terraform
/
```

### Crash Output

```
Initializing modules...

Initializing the backend...

Initializing provider plugins...
- Finding latest version of hashicorp/databricks...
- Reusing previous version of hashicorp/azurerm from the dependency lock file
- Reusing previous version of databrickslabs/databricks from the dependency lock file
- Reusing previous version of hashicorp/random from the dependency lock file
- Using previously-installed hashicorp/azurerm v2.51.0
- Using previously-installed databrickslabs/databricks v0.3.1
- Using previously-installed hashicorp/random v3.1.0

Error: Failed to query available provider packages

Could not retrieve the list of available versions for provider
hashicorp/databricks: provider registry registry.terraform.io does not have a
provider named registry.terraform.io/hashicorp/databricks

If you have just upgraded directly from Terraform v0.12 to Terraform v0.14
then please upgrade to Terraform v0.13 first and follow the upgrade guide for
that release, which might help you address this problem.
```

### Expected Behavior

I have a `main.tf` file which includes my providers section as followed: 

```terraform
terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "2.51.0"
    }

    databricks = {
      source = "databrickslabs/databricks"
      version = "0.3.1"
    }

  }

 ....
  }
}

#---------------- PROVIDERS CONFIG ----------------#
provider "azurerm" {
  features {}
  skip_provider_registration = true
}


provider "databricks" {
  azure_workspace_resource_id = module.azurerm_databricks_workspace.databricks_id

}

module "azurerm_databricks_workspace" {
  source = "../../_modules/azurerm_databricks_workspace"

  client_short       = var.client_short
  environment_suffix = var.environment_suffix
  resource_group     = var.resource_group
  tags               = var.default_tags
}

module "databricks_cluster" {
  source = "../../_modules/databricks_cluster"
  
}
```
Should result in using `registry.terraform.io/databrickslabs/databricks` in the module.

### Actual Behavior

but it results in 

```
.
├── provider[registry.terraform.io/databrickslabs/databricks] 0.3.1
├── provider[registry.terraform.io/hashicorp/azurerm] 2.51.0
...
├── module.azurerm_databricks_workspace
│   ├── provider[registry.terraform.io/hashicorp/random]
│   └── provider[registry.terraform.io/hashicorp/azurerm]
├── module.databricks_cluster
│   └── provider[registry.terraform.io/hashicorp/databricks]         <-------------- ?
...
```

### Steps to Reproduce

1. main.tf with providers section databricks 
2. create module that tries to use this provider

### Additional Context

When I explicitly add the required providers databricks section inside the module it is able to use the correct provider.

```
.
├── provider[registry.terraform.io/hashicorp/azurerm] 2.51.0
├── provider[registry.terraform.io/databrickslabs/databricks] 0.3.1
├── module.azurerm_databricks_workspace
│   ├── provider[registry.terraform.io/hashicorp/random]
│   └── provider[registry.terraform.io/hashicorp/azurerm]
├── module.databricks_cluster
│   └── provider[registry.terraform.io/databrickslabs/databricks] 0.3.1
...
```

**Comment by jbardin** [comment]
Hi @dsfrederic 

Provider _configuration_ can be passed between modules, but the _source_ of the provider is never inherited, and must be declared within the module. When there is no source declared, the default namespace is used, which is where the `provider[registry.terraform.io/hashicorp/databricks]` provider name comes from.

Passing in a configuration like you did should not have worked due to the lack of `required_providers`, which is fixed in the 0.15 release, and will result in an error like:

```
╷
│ Error: Invalid type for provider module.mod.provider["registry.terraform.io/hashicorp/null"]
│
│   on main.tf line 15, in module "azurerm_databricks_workspace":
│   15:     databricks = databricks
│
│ Cannot use configuration from provider[registry.terraform.io/databrickslabs/databricks] for 
| module.azurerm_databricks_workspace.provider[registry.terraform.io/databrickslabs/databricks].
| The given provider configuration is for a different provider type.
╵
```

Thanks!

**Comment by dsfrederic** [comment]
@jbardin so I can conclude I'll have to add the `required_providers` section to every module. Do I have to **explicitly** pass the configuration to every module? Like this:

```
provider "databricks" {
  alias = "databricks_provider"

  azure_workspace_resource_id = module.azurerm_databricks_workspace.databricks_id
}

module "name" {
  source = "../../_modules/databricks_cluster"
  providers = {
    databricks = databricks.databricks_provider
  }
}
```

What exactly will be fixed in 0.15? Is it the mismatch between a provider configuration and the configured provider inside of a module? 

Am I right to say that for azurerm this wasn't an issue because this is part of the default namespace (`registry.terraform.io/hashicorp`) ?

**Comment by jbardin** [comment]
Yes, all providers outside of the default `registry.terraform.io/hashicorp` must be specified in a `required_providers` block within the module. If you want to use an aliased provider name, you must pass it into the module explicitly. The azurerm provider was working without declaring it because it is part of the `registry.terraform.io/hashicorp` namespace. 

0.15 will contain more validation to prevent incorrect usage of provider configurations, however the fact that an undeclared provider will default to the `registry.terraform.io/hashicorp` namespace is unchanged, as that was a deliberate design decision. 

We use GitHub issues for tracking bugs and enhancements, rather than for questions. While we can sometimes help with certain simple problems here, it's better to use [the community forum](https://discuss.hashicorp.com/c/terraform-core) where there are more people ready to help.

Thanks!
